src/test/java/org/folio/circulation/support/utils/DateTimeUtilTests.java

```diff
@@ -0,0 +1,461 @@
+package org.folio.circulation.support.utils;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import java.time.LocalDate;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+
+class DateTimeUtilsTests {
+
+  @ParameterizedTest
+  @MethodSource("shouldGetCorrectStartOfYearForZonedDateTimeParameters")
+  void shouldGetCorrectStartOfYearForZonedDateTime(int id, ZonedDateTime from, ZonedDateTime expected, ZoneId zone) {
+    final ZonedDateTime result = zone == null
+      ? DateTimeUtil.atStartOfYear(from)
+      : DateTimeUtil.atStartOfYear(from, zone);
+
+    assertEquals(expected.format(DateTimeFormatter.ISO_DATE), result.format(DateTimeFormatter.ISO_DATE), "For test " + id);
+  }
+
+  private static Object[] shouldGetCorrectStartOfYearForZonedDateTimeParameters() {
+    return new Object[] {
+      new Object[] { 0, ZonedDateTime.parse("2018-03-01T02:28:19-05:00"), ZonedDateTime.parse("2018-01-01T00:00:00-05:00"), null },
+      new Object[] { 1, ZonedDateTime.parse("2018-02-28T21:28:19+00:00"), ZonedDateTime.parse("2018-01-01T00:00:00-05:00"), ZoneId.of("America/New_York") },
+      new Object[] { 2, ZonedDateTime.parse("2018-01-01T00:14:03-05:00"), ZonedDateTime.parse("2018-01-01T00:00:00-05:00"), null },
+      new Object[] { 3, ZonedDateTime.parse("2018-01-01T00:14:03+00:00"), ZonedDateTime.parse("2017-01-01T00:00:00-05:00"), ZoneId.of("America/New_York") },
+      new Object[] { 4, ZonedDateTime.parse("2018-12-31T23:40:50+01:00"), ZonedDateTime.parse("2018-01-01T00:00:00+01:00"), null },
+      new Object[] { 5, ZonedDateTime.parse("2018-12-31T23:40:50+00:00"), ZonedDateTime.parse("2019-01-01T00:00:00+01:00"), ZoneId.of("Europe/Warsaw") }
```

Comment:
This is a short list of areas where I believe if something could go wrong these are the most likely places.
0 = A date somewhere between the start and end of the year, picking the dates.
1 = The same date as above but in UTC time zone, then have the time zone applied.
2 = January 1st 2018 (which in UTC is december in 2017), feels like a good test to confirm we get 2018 in the desired time zone.
3 = Similar to above but starting with a UTC date. Given that Jan 1 2018 in UTC (near midnight) could be Dec 31 2017 in a different time zone, we need to get start of year in 2017 because that is the year relative to the time zone (-5).
4 = Same concept as 2, but for the end of the year.
5 = Same concept as 3, but for the end of the year.

The even numbers (including 0) represent the `DateTimeUtil.atStartOfYear()` call without the time zone passed and so the time zone provide is what is used for the calculation.
The odd numbers represents `DateTimeUtil.atStartOfYear()` with a time zone passed, so the provided date is changed to the requested time zone and then the start of the year is calculated.

This logic is re-used for each of the respectice start/end of year, month, day, etc.. methods.
The specific times are changed to test notable points of failure, such as the start or end of an the year, month, day, etc..

To keep the tests short, I chose a middle of the range, the start of the range, and the end of th range for a broad sweep testing process.

Reply:
Thank you for explaining that to me.

My concern with this style of tests is that I don't know how someone coming to this test in future would understand the reasoning or distinction between these examples.

(similar to the conversation we were having elsewhere in this review).